### PR TITLE
Coupons: Remove redundant animation when reloading coupon list

### LINKED ISSUE #6796: NSInternalInconsistencyException: Fatal: supplied item identifiers are not unique. Duplicate identifiers: {(
Sentry Issue: [WOOCOMMERCE-IOS-1DCN](https://sentry.io/organizations/a8c/issues/3252194625/?referrer=github_integration)

```
NSInternalInconsistencyException: Fatal: supplied item identifiers are not unique. Duplicate identifiers: {(
    WooCommerce.TitleAndSubtitleAndStatusTableViewCell.ViewModel(title: "", subtitle: "0,00€ sur 1 Produit", accessibilityLabel: "", status: "Expiré", statusBackgroundColor: <UIDynamicCatalogColor: 0x280765a90; name = Gray5>)
)}
  File "CouponListViewController.swift", line 141, in CouponListViewController.configureViewModel
  File "<compiler-generated>", in thunk for closure
  File "<compiler-generated>", in thunk for closure
  File "Classes/Model/Coupon+Woo.swift", in CouponListViewModel.couponViewModels.setter
  File "CouponListViewModel.swift", line 81, in CouponListViewModel.buildCouponViewModels
...
(48 additional frame(s) were not displayed)
```

### LINKED ISSUE #7135: Coupon: Odd animations in items of coupon list
# Problem
When loading a coupon list, there's an obvious animation in the items when the list is loaded. This also happens when the list is reloaded.

# Steps to reproduce
- Make sure that you have Coupon Management enabled in Settings > Experimental Features.
- Navigate to Menu > Coupons.
- If the list contains at least 2 items, notice that there's an odd animation whenever the list is loaded.

# Screenshot

https://user-images.githubusercontent.com/5533851/175209589-e9159e82-848a-41ed-8853-eda4552d22d5.mp4

# Expectation
The redundant animation should be removed.

### PR DESCRIPTION
**Pull request by itsmeichigo**
Closes: #7135 

### Description

When trying to fix #6796, I introduced a new unique identifier to `TitleAndSubtitleAndStatusTableViewCell`'s view model to avoid sending duplicating items to the coupon list diffable table view data source. Since the identifier was of type `UUID`, every time a cell view model is created, it's a brand new one to the data source and therefore causes the list to animate when a new snapshot is applied.

This PR fixes the above issue with 2 parts:
- Make the cell view model identifier a string and use coupon IDs to construct the identifiers. This is to make sure that each cell is identified by the coupon it is displaying and stop the cells from animating when the list is reloaded.
- For some reason with the above change in place, the list still jumps a bit when reloaded. It looks like the offset of the list is always reset to show the top item. To fix this, I added `applySnapshotUsingReloadData` to reload the UI without animation. This function is only available since iOS 15, so the problem persists on iOS 14.

### Testing instructions

- Make sure that you have Coupon Management enabled in Settings > Experimental Features.
- Navigate to Menu > Coupons. If your list has more than 2 items, notice that the list doesn't jump when it is reloaded. Try pulling to refresh the list to confirm this.

### Screenshots

https://user-images.githubusercontent.com/5533851/175230562-d93509dc-ed0c-4b37-ace7-862356d7169c.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/ReusableViews/TitleAndSubtitleAndStatusTableViewCell.swift [hunk 1]
```diff
@@ -37,8 +37,7 @@ final class TitleAndSubtitleAndStatusTableViewCell: UITableViewCell, SearchResul
 extension TitleAndSubtitleAndStatusTableViewCell {
     struct ViewModel: Hashable {
         /// A unique ID to avoid duplicated identifier for the view model in diffable datasource
-        private let id = UUID()
-
+        var id: String = UUID().uuidString
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/ReusableViews/TitleAndSubtitleAndStatusTableViewCell.swift:42]: Nit: I think we don't have to explicitly specify the type here.

File: WooCommerce/Classes/ViewRelated/ReusableViews/TitleAndSubtitleAndStatusTableViewCell.swift [hunk 2]
```diff
@@ -37,8 +37,7 @@ final class TitleAndSubtitleAndStatusTableViewCell: UITableViewCell, SearchResul
 extension TitleAndSubtitleAndStatusTableViewCell {
     struct ViewModel: Hashable {
         /// A unique ID to avoid duplicated identifier for the view model in diffable datasource
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/ReusableViews/TitleAndSubtitleAndStatusTableViewCell.swift]: Could we update the comment to explain why `id` is `var` and when to override the default value? 

Do we have to use our own unique IDs whenever we use diffable data sources? 🤔

File: WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift [hunk 3]
```diff
@@ -153,7 +153,13 @@ final class CouponListViewController: UIViewController, GhostableViewController
                 var snapshot = NSDiffableDataSourceSnapshot<Section, CouponListViewModel.CellViewModel>()
                 snapshot.appendSections([.main])
                 snapshot.appendItems(viewModels, toSection: Section.main)
-                self.dataSource.apply(snapshot)
+
+                if #available(iOS 15.0, *) {
+                    // minimally reloads the list without computing diff or animation
+                    self.dataSource.applySnapshotUsingReloadData(snapshot)
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift:159]: - I am new to this `applySnapshotUsingReloadData` method. I hope that this avoids calculating diff and reloads the view directly. If we are trying to apply the snapshot without the animation, could you kindly explain why we can't use `self.dataSource.apply(snapshot, animatingDifferences: false)` instead? 

- Also, I tested in `iPhone 13 Pro Max` running `iOS 15.4.1` and I couldn't notice any problems with animation while just using `self.dataSource.apply(snapshot)`. i.e. The other changes from this PR related to using `couponID` as `id` fixed the animation issue for me. 🤔

**Review by selanthiraiyan — APPROVED** [review]
Thanks for fixing this, Huong. 🎉 

I left a couple of questions. 😀

File: WooCommerce/Classes/ViewRelated/ReusableViews/TitleAndSubtitleAndStatusTableViewCell.swift (hunk 1 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/ReusableViews/TitleAndSubtitleAndStatusTableViewCell.swift:42]: I'd like to keep the type declaration here to be consistent with the other properties 😅 Also it's clearer to see the type than having to interpret the default value (it's the same for the compiler as well). Does that make sense?

File: WooCommerce/Classes/ViewRelated/ReusableViews/TitleAndSubtitleAndStatusTableViewCell.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/ReusableViews/TitleAndSubtitleAndStatusTableViewCell.swift]: We should define the ID for more efficient loading when using diffable data sources. I updated the comment in 9a129884a16eaeac052fce04d4985f4f07db258d, thanks!

File: WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift (hunk 3 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift:159]: Is it possible that you have dismissed the header banner of the coupon list? I'm still keeping mine so using either `apply(snapshot)` or `apply(snapshot, animatingDifferences: false)` I still see the animation for changing the table view offset (I'm running on iOS 15.5 simulator):


https://user-images.githubusercontent.com/5533851/176080787-920bde11-75e8-45ab-9ea5-5352048cc803.mp4

> I am new to this `applySnapshotUsingReloadData` method. I hope that this avoids calculating diff and reloads the view directly. If we are trying to apply the snapshot without the animation, could you kindly explain why we can't use self.dataSource.apply(snapshot, animatingDifferences: false) instead?

Looking at the documentation of the two methods, it seems that `applySnapshotUsingReloadData` doesn't compute diff whereas `apply` does and has the option to not animate any changes. The documentation doesn't mention the change in the offset of the table view though, so I don't have a good answer for why the former doesn't cause the animation in the video above 😅

Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponListViewController.swift:159]: > I still see the animation for changing the table view offset

I didn't have enough coupons for "table view offset change" to happen. It makes sense to me now. Thanks for the details.

**Comment by selanthiraiyan** [comment]
Thanks for the answers, Huong. 😁 

:shipit:

**Pull request merged**